Port Poseidon to latest canonical + microkelvin + nstack

CPerezz opened the pull request:
Resolves: #125

Commit:
Update tmp deps

Commit:
Replace rand by rand_core

Resolves: #127.

Commit:
Rework the trait system for the crate structs

Most of the crate structures changed it's requirements since there
were changes in the traits that were implemented for them or on which
they were relying on.

Therefore `PoseidonAnnotation`, `PoseidonLeaf` and `PoseidonMaxAnnotation`
have been updated

Commit:
Update cipher circuits to lates plonk version changes

Commit:
Fix PoseidonMaxAnnotation

- Remove `PoseidonWalkableAnnotation` since is no longer
needed.
- Implement `Combine` for `PoseidonMaxAnnotation`

Commit:
Re-implement from_generic_node more generically

This needed to take a generic `Compound` structure implemented
for which we can extract the Cardinality.

This goes inline also with the update of the deps and the new APIs
that they have`

Commit:
Refactor public API for `PoseidonTree`

- Don't require a specific `Store` instance in the type.
- Update get signature to get `u64`.
- Update `push` to avoid major computations to obtain curren
size and simplify the code. Also, use `u64`.
- Update `branch` method to use `u64`.
- Remove any `PoseidonIterator` instance  existance and impl
`iter_walk`.

Commit:
Fix doc examples with latest lib changes

Commit:
Fix all warnings for the repo

vlopes11 commented on Cargo.toml:
```diff
@@ -72,3 +68,7 @@ panic = 'abort'
+
+[patch.crates-io]
+nstack = {path = "../nstack", optional = true}
+dusk-hades = { path = "../Hades252", default-features = false }
```
Used local path + line break

vlopes11 commented on src/tree/tree.rs:
```diff
@@ -131,100 +108,21 @@ where
-    pub fn iter_walk<D: Clone>(
+    /// Provides an iterator over the leaves of the tree from a provided starting point.
+    /// To iterate the entire tree, simply provide `0` as `start`.
+    pub fn iter_walk(
```
This change looks like a regression. We had a more powerful iterator that could iter from generic criteria and we reduced it for an incremented index. Maybe we just want to create a different method?

CPerezz replied:
You're right!. It has been added back in 98e9d2c

vlopes11 commented on src/tree/tree.rs:
```diff
@@ -131,100 +108,21 @@ where
+        let result = self.inner.nth(start);
+        match result {
+            Ok(Some(iter)) => Ok(iter),
+            _ => Err(CanonError::NotFound),
```
No differentiation between an actual error in the backend or `CanonError::NotFound`

CPerezz replied:
Not sure I follow

vlopes11 commented on src/tree/mod.rs:
````diff
@@ -10,15 +10,14 @@
 //! ### Example
 //!
-//! ```rust
+//! ```rust,no_run
````
Why documentation tests are removed? Its important to keep its consistency

CPerezz replied:
They are not removed. This flag makes forces the compiler to compile the tests but not run them. And considering that what it done in the test takes quite some time, I think it's Ok to just see if it compiles but not run it.

vlopes11 reviewed (changes requested):
Few nts

Commit:
Fix imports

Commit:
Implement custom walkable fn for PoseidonTree

Commit:
Remove no_run tags from doc-tests

krl commented on src/tree/tree.rs:
```diff
@@ -125,4 +125,20 @@ where
+    /// This is only useful if annotate the tree is going to make the iteration perform sub-linearly.
+    pub fn annotated_iter_walk(
+        &self,
+        walker: impl Walker<NStack<L, A>, A>,
```
i would go for a <W> where W: Walker on this, using impl in argument position is a bit controversial.

CPerezz replied:
Makes sense. Changed!

Commit:
Refactor error module

- Removes unused std-compatibility for the `Error` struct.
- Removes the usage of `anyhow` and `thiserror` in
tests and lib.
- Adds two new variants for the `Error` called `TreeIterFailed`
and `TreeBranchFailed`.

Resolves: #132

Commit:
Add Walker trait bounds insinde where clause

Commit:
Fix comments

Commit:
Remove unnecessary cfg flags

Commit:
Update CHANGELOG.md

CPerezz committed:
Merge branch 'canon_microkelvin' into remove_anyhow

CPerezz committed:
Merge pull request #133 from dusk-network/remove_anyhow

Refactor error module & remove anyhow + thiserror

ZER0 commented on Cargo.toml:
```diff
@@ -72,3 +64,8 @@ panic = 'abort'
+[patch.crates-io]
+nstack = {path = "../nstack", optional = true}
+dusk-hades = { path = "../Hades252", default-features = false }
+microkelvin = {git = "https://github.com/dusk-network/microkelvin", branch = "expose_walker"}
```
Add new line

Commit:
tmp

Commit:
Make PoseidonTreeAnnotation an auto-trait

Commit:
Implement example for custom walker iterators

Commit:
Merge release-0.21 into canon_microkelvin

Commit:
Move tests as integration tests when internals are not used

Resolves: #134

Commit:
Use final versions of deps

Commit:
Merge branch 'update_deps' into rand_core

Commit:
Merge branch 'rand_core' into canon_microkelvin

Commit:
Remove canon_host-related checks from CI

Resolves: #136

Commit:
Update rustfmt toolchain version

Commit:
Remove anyhow leftovers from docs

Commit:
Use alloc feature by default for plonk & hades

Commit:
Change the feature requirements of the crate

The requirements were not in-line with the dependencies that we
are using and it's feature sets.

Therefore, we no longer need to restrict the zk part of things
to `std` and instead can always have it enabled via using `alloc`
by default.

Resolves: #138

Commit:
Fix header indentation for dusk-analyzer

Commit:
Remove unnecessary cfg std flag

ZER0 commented on CHANGELOG.md:
```diff
@@ -7,6 +7,12 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ## [Unreleased]
 
+### Changed
+- Updated `error` module to be simpler and no_std friendly. [#132](https://github.com/dusk-network/poseidon252/issues/132)
```
1. Present tense: `Update` not `Updated`
2. No period at the end of the sentence (so `friendly` not `friendly.`

You can remove the "simpler", ideally if you notice you have to put `and` maybe you're either over-complicate or adding two separate thing into one. The simplification can be considered as byproduct of the no_std. So I would actually change into:
```
- Change `error` module to be `no_std` compatible
```
Or something like that.

ZER0 commented on CHANGELOG.md:
```diff
@@ -7,6 +7,12 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
+- Updated `error` module to be simpler and no_std friendly. [#132](https://github.com/dusk-network/poseidon252/issues/132)
+
+### Removed
+- Removed `anyhow` and `thiserror` from deps. [#132](https://github.com/dusk-network/poseidon252/issues/132)
```
- Past tense: `Remove` not `Removed`
- No period at the end of sentence (so `deps` not `deps.`)

Commit:
Update CHANGELOG.md

Commit:
Rollback fmt rustc version to nightly-2020-10-25

ZER0 commented on README.md:
```diff
@@ -119,11 +118,11 @@ impl PoseidonLeaf<MemStore> for DataLeaf {
 
-    // Instantiate a new tree with the MemStore implementation
-    let mut tree: PoseidonTree<DataLeaf, PoseidonAnnotation, MemStore, DEPTH> =
+    // Instantiate a new tree.
```
No need for period (the other comments doesn't have it, so prefer consistency)

ZER0 commented on src/error.rs:
```diff
@@ -4,39 +4,27 @@
+pub enum Error {
     /// Error pushing to the poseidon tree
-    TreePushFailed(E),
+    TreePushFailed,
     /// Error on pop of the tree
-    TreePopFailed(E),
+    TreePopFailed,
     /// Error fetching the Nth item from the tree
-    TreeGetFailed(E),
+    TreeGetFailed,
+    /// Failed to obtain a Branch from a tree.
+    TreeBranchFailed,
+    /// Failed to obtain an Iterator from a tree.
+    TreeIterFailed,
     /// Decryption failed for the provided secret+nonce
     CipherDecryptionFailed,
 }
```
Wondering why this peculiar naming convention: why add `Failed` at suffix? It's not very idiomatic. Not suggesting to change now at all (it would be a MASSIVE breaking change) it just looks a bit weird. Maybe we could change before the release `1.0`?

CPerezz replied:
Filed https://github.com/dusk-network/Poseidon252/issues/140

ZER0 commented on src/tree/tests.rs:
```diff
@@ -7,21 +7,28 @@
+}
+
 impl MockLeaf {
     pub fn random<R: RngCore + CryptoRng>(rng: &mut R) -> Self {
```
If it's never used as the CI is saying, and never exported since it's a test, should be removed? Or it's used in the later PRs?

CPerezz replied:
It's used later and the CI doesn't give anymore warnings.

Commit:
Remove redundant mod

ZER0 reviewed (approved):
LGTM, but address the comments first, and rephrase the commit's messae to ensure they're folliwing the guidelines (e.g. starting with a changelog's verb etc)

CPerezz committed:
Merge pull request #135 from dusk-network/integration_tests

Move tests as integration tests whenever possible

vlopes11 reviewed (approved):
LGTM

Commit:
Update CHANGELOG.md
